homeassistant/components/sensor/ebox.py

```diff
@@ -0,0 +1,306 @@
+    'upload': ['Upload',
+               GIGABITS, 'mdi:upload'],
+    'total': ['Total',
+              GIGABITS, 'mdi:download'],
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_MONITORED_VARIABLES):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+HOST = "https://client.ebox.ca"
+HOME_URL = "{}/".format(HOST)
+LOGIN_URL = "{}/login".format(HOST)
+USAGE_URL = "{}/myusage".format(HOST)
+
+USAGE_MAP = {"before_offpeak_download": 0,
+             "before_offpeak_upload": 1,
+             "before_offpeak_total": 2,
+             "offpeak_download": 3,
+             "offpeak_upload": 4,
+             "offpeak_total": 5,
+             "download": 6,
+             "upload": 7,
+             "total": 8}
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the EBox sensor."""
+    # Create a data fetcher to support all of the configured sensors. Then make
+    # the first call to init the data.
+
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    try:
+        ebox_data = EBoxData(username, password)
+        ebox_data.update()
+    except requests.exceptions.HTTPError as error:
+        _LOGGER.error(error)
+        return False
+
+    name = config.get(CONF_NAME)
+
+    sensors = []
+    for variable in config[CONF_MONITORED_VARIABLES]:
+        sensors.append(EBoxSensor(ebox_data, variable, name))
+
+    add_devices(sensors)
+
+
+class EBoxSensor(Entity):
+    """Implementation of a EBox sensor."""
+
+    def __init__(self, ebox_data, sensor_type, name):
+        """Initialize the sensor."""
+        self.client_name = name
+        self.type = sensor_type
+        self.entity_id = "sensor.{}_{}".format(name, sensor_type)
+        self._name = SENSOR_TYPES[sensor_type][0]
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+        self._icon = SENSOR_TYPES[sensor_type][2]
+        self.ebox_data = ebox_data
+        self._state = None
+
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return '{} {}'.format(self.client_name, self._name)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    def update(self):
+        """Get the latest data from EBox and update the state."""
+        self.ebox_data.update()
+        if self.type in self.ebox_data.data:
+            self._state = round(self.ebox_data.data[self.type], 2)
+
+
+class EBoxData(object):
```

Comment:
This platform contains too much protocol specific details. Please extract into a 3rd party lib and base it upon that.